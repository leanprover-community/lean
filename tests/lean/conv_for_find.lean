-- typos should be caught
example {a b c : ℕ} : (a * b) * c = (a * b) * c := begin
  conv {
    rw does_not_exist,
  },
end

example {a b c : ℕ} : (a * b) * c = (a * b) * c := begin
  conv {
    for (_ * _) [1] {
      rw does_not_exist,
    },
  },
end

example {a b c : ℕ} : (a * b) * c = (a * b) * c := begin
  conv {
    find (_ * _) {
      rw does_not_exist,
    },
  },
end

-- filled in metavariables should remain filled
example {a b : ℕ} : ∃ x, x * a = a * x :=  begin
  existsi _,
  conv {
    rw nat.mul_comm b a,
  },
end

-- but this unfortunately doesn't work for `find` and `for`
example {a b : ℕ} : ∃ x, x * a = a * x :=  begin
  existsi _,
  conv {
    find (_ * _) {
      rw nat.mul_comm b a,
    },
  },
end

example {a b : ℕ} : ∃ x, x * a = a * x :=  begin
  existsi _,
  conv {
    for (_ * _) [1] {
      rw nat.mul_comm b a,
    },
  },
end

-- matching should work under binders
example {ι} (p : ι → Prop) : (∀ i, p i) ↔ (∀ j, p j) :=
begin
  conv {
    find (p _) {
      trace_lhs,
    },
  },
end

example {ι} (p : ι → Prop) : (∀ i, p i) ↔ (∀ j, p j) :=
begin
  conv {
    for (p _) [1, 2] {
      trace_lhs,
    },
  },
end
