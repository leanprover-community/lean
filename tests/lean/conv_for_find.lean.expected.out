conv_for_find.lean:4:4: error: unknown identifier 'does_not_exist'
state:
a b c : ℕ
⊢ a * b * c = a * b * c = ?m_1
conv_for_find.lean:11:6: error: unknown identifier 'does_not_exist'
state:
a b c : ℕ
⊢ a * b * c = ?m_1
conv_for_find.lean:19:6: error: unknown identifier 'does_not_exist'
state:
a b c : ℕ
⊢ a * b * c = ?m_1
conv_for_find.lean:40:0: error: tactic failed, there are unsolved goals
state:
a b : ℕ
⊢ ℕ
conv_for_find.lean:49:0: error: tactic failed, there are unsolved goals
state:
a b : ℕ
⊢ ℕ
p i
p i
p j
